src/jrd/btr.cpp

```diff
@@ -679,152 +682,161 @@ void BTR_evaluate(thread_db* tdbb, const IndexRetrieval* retrieval, RecordBitmap
+				prefix = 0;
+			}
 		}
-
-		while (true)
+		else
 		{
-			if (node.isEndLevel) {
-				break;
+			// if there isn't an upper bound, just walk the index to the end of the level
+			const UCHAR* endPointer = (UCHAR*)page + page->btr_length;
+			const bool ignoreNulls =
+				(retrieval->irb_generic & irb_ignore_null_value_key) && (idx.idx_count == 1);
+
+			IndexNode node;
+			pointer = node.readNode(pointer, true);
+			// Check if pointer is still valid
+			if (pointer > endPointer) {
+				BUGCHECK(204);	// msg 204 index inconsistent
 			}
 
-			if (!node.isEndBucket)
+			while (true)
 			{
-				// If we're walking in a descending index and we need to ignore NULLs
-				// then stop at the first NULL we see (only for single segment!)
-				if (descending && ignoreNulls && node.prefix == 0 &&
-					node.length >= 1 && node.data[0] == 255)
-				{
+				if (node.isEndLevel) {
 					break;
 				}
 
-				if (skipLowerKey)
+				if (!node.isEndBucket)
 				{
-					checkForLowerKeySkip(skipLowerKey, partLower, node, lower, idx, retrieval);
-				}
+					// If we're walking in a descending index and we need to ignore NULLs
+					// then stop at the first NULL we see (only for single segment!)
+					if (descending && ignoreNulls && node.prefix == 0 &&
+						node.length >= 1 && node.data[0] == 255)
+					{
+						break;
+					}
 
-				if (!skipLowerKey)
-				{
-					if (!bitmap_and || bitmap_and->test(node.recordNumber.getValue()))
-						RBM_SET(tdbb->getDefaultPool(), bitmap, node.recordNumber.getValue());
+					if (skipLowerKey)
+					{
+						checkForLowerKeySkip(skipLowerKey, partLower, node, *lower, idx, retrieval);
+					}
+
+					if (!skipLowerKey)
+					{
+						if (!bitmap_and || bitmap_and->test(node.recordNumber.getValue()))
+							RBM_SET(tdbb->getDefaultPool(), bitmap, node.recordNumber.getValue());
+					}
+					pointer = node.readNode(pointer, true);
+					// Check if pointer is still valid
+					if (pointer > endPointer) {
+						BUGCHECK(204);	// msg 204 index inconsistent
+					}
+					continue;
 				}
+
+				page = (btree_page*) CCH_HANDOFF(tdbb, &window, page->btr_sibling, LCK_read, pag_index);
+				endPointer = (UCHAR*) page + page->btr_length;
+				pointer = page->btr_nodes + page->btr_jump_size;
 				pointer = node.readNode(pointer, true);
 				// Check if pointer is still valid
 				if (pointer > endPointer) {
 					BUGCHECK(204);	// msg 204 index inconsistent
 				}
-				continue;
-			}
-
-			page = (btree_page*) CCH_HANDOFF(tdbb, &window, page->btr_sibling, LCK_read, pag_index);
-			endPointer = (UCHAR*) page + page->btr_length;
-			pointer = page->btr_nodes + page->btr_jump_size;
-			pointer = node.readNode(pointer, true);
-			// Check if pointer is still valid
-			if (pointer > endPointer) {
-				BUGCHECK(204);	// msg 204 index inconsistent
 			}
 		}
-	}
 
-	CCH_RELEASE(tdbb, &window);
+		CCH_RELEASE(tdbb, &window);
+	} while ((lower = lower->key_next.get()) && (upper = upper->key_next.get()));
```

Comment:
I'm not sure putting the whole body into the loop is a good thing. First of all, the parts with `retrieval->irb_lower_count == 0` and `retrieval->irb_upper_count == 0` will never be executed more than once, as `irb_multi_starting` implies that both bounds are defined. Second, it's worth testing that locating the next lower bound via `BTR_find_page()` (i.e. from the tree root) is really faster than iterating through `find_node_start_point()` (i.e. using jump nodes and sibling pointers) (*). I expect contractions to have a common prefix (e.g. 'C' and 'CH'), so they should be located quite close to each other on leaf pages.

I don't request these changes, just want you to consider them, at least for master.

(*) I mean something like this:

```c++
if (retrieval->irb_upper_count)
{
  while (true)
  {
    while (scan(tdbb, pointer, bitmap, bitmap_and, &idx, retrieval, prefix, upper,
                       skipLowerKey, *lower))
    {
      page = (btree_page*) CCH_HANDOFF(tdbb, &window, page->btr_sibling, LCK_read, pag_index);
      pointer = page->btr_nodes + page->btr_jump_size;
      prefix = 0;
    }

    if ((!(lower = lower->key_next.get()) && (upper = upper->key_next.get())))
      break;

    while (!(pointer = find_node_start_point(page, lower, 0, &prefix,
              idx.idx_flags & idx_descending, (retrieval->irb_generic & (irb_starting | irb_partial)))))
    {
      page = (btree_page*) CCH_HANDOFF(tdbb, &window, page->btr_sibling, LCK_read, pag_index);
    }

    prefix = IndexNode::computePrefix(upper->key_data, upper->key_length,
                                                            lower->key_data, lower->key_length);
  }
}
```

Reply:
> I expect contractions to have a common prefix (e.g. 'C' and 'CH'), so they should be located quite close to each other on leaf pages.

No. CH sorts after H while C sorts after B.

Reply:
That's sad ;-)

Reply:
Could your approach have degraded performance due to this fact (distance of keys)?

Reply:
Yes, but it's mostly targeted at many keys being scanned at once, as the more keys are present the more a range scan (via sibling pointers) wins over multiple lookups from the tree root. For the case of few keys I just fallback to the old algorithm (independent lookups OR'ed into a single bitmap), as I deal with equalities only. I believe I will be able to utilize your solution for this case.